### PR TITLE
Add unique ID maps and engine

### PR DESCRIPTION
**Pull request by peachbits**

### CODE REVIEW

File: src/rates.ts [hunk 1]
```diff
@@ -143,18 +150,18 @@ export const getExchangeRates = async (
   localDb: any
 ): Promise<ReturnGetRate> => {
   try {
-    const dates: string[] = []
+    const docs: string[] = ['currencyCodeMaps']
```
Comment (peachbits) [inline src/rates.ts:153]: Don't get with every call. Memoize instead

File: src/providers/coinMarketCap.ts [hunk 2]
```diff
@@ -33,6 +31,16 @@ const {
   defaultFiatCode: DEFAULT_FIAT
 } = config
 
+export const hasUniqueId = (code: string, assetMap: AssetMap): boolean =>
+  assetMap[code] != null
+
+export const invertCodeMapKey = (id: string, assetMap: AssetMap): string => {
```
Comment (peachbits) [inline src/providers/coinMarketCap.ts:36]: one-liner

File: src/providers/coinMarketCap.ts [hunk 3]
```diff
@@ -180,20 +191,24 @@ export const coinMarketCap = async (
...
   for (const pair of rateObj) {
-    if (datesAndCodesWanted[pair.date] == null) {
-      datesAndCodesWanted[pair.date] = [...DEFAULT_CODES]
-    }
     const fromCurrency = checkConstantCode(fromCode(pair.currency_pair))
-    if (!isFiatCode(fromCurrency) && !DEFAULT_CODES.includes(fromCurrency)) {
-      datesAndCodesWanted[pair.date].push(fromCurrency)
+    if (
+      !isFiatCode(fromCurrency) &&
+      !DEFAULT_CODES.includes(fromCurrency) &&
```
Comment (peachbits) [inline src/providers/coinMarketCap.ts:196]: remove this

File: src/rates.ts [hunk 4]
```diff
@@ -189,13 +189,25 @@ export const getExchangeRates = async (
 }
 
 export const currencyBridgeDB = (rateObj: ReturnGetRate): void => {
+  let constantCurrencyCodes = {}
```
Comment (peachbits) [inline src/rates.ts:192]: Pass relevant asset map

File: src/utils/dbUtils.ts [hunk 5]
```diff
@@ -1,45 +1,34 @@
...
-  locks[LOCK_ID].acquireAsync().then(() =>
-    localDB
-      .bulk({ docs })
-      .then(response => {
-        const successArray = response
-          .filter(doc => doc.error == null)
-          .map(doc => doc.id)
-        if (successArray.length > 0)
-          logger(`Saved document IDs: ${successArray.join(', ')} to db: ${db}`)
+  localDB
```
Comment (peachbits) [inline src/utils/dbUtils.ts:10]: add simple rev based conflict resolution. Keep highest

File: package.json [hunk 6]
```diff
@@ -58,7 +58,6 @@
     "typescript": "^3.8.2"
   },
```
Comment (thehobbit85) [inline package.json:59]: Change!

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
Changes!

**Review by thehobbit85 — CHANGES_REQUESTED** [review]

**Pull request merged**